backend/danswer/direct_qa/__init__.py

```diff
@@ -43,24 +47,52 @@ def check_model_api_key_is_valid(model_api_key: str) -> bool:
 
 def get_default_backend_qa_model(
     internal_model: str = INTERNAL_MODEL_VERSION,
-    api_key: str | None = None,
+    endpoint: str | None = GEN_AI_ENDPOINT,
+    model_host_type: str | None = GEN_AI_HOST_TYPE,
+    api_key: str | None = get_gen_ai_api_key(),
     timeout: int = QA_TIMEOUT,
     **kwargs: Any
 ) -> QAModel:
-    if internal_model == "openai-completion":
+    if internal_model in [
+        DanswerGenAIModel.GPT4ALL.value,
+        DanswerGenAIModel.GPT4ALL_CHAT.value,
+    ]:
+        # gpt4all is not compatible M1 Mac hardware as of Aug 2023
+        pkg_resources.get_distribution("gpt4all")
+
+    if internal_model == DanswerGenAIModel.OPENAI.value:
         return OpenAICompletionQA(timeout=timeout, api_key=api_key, **kwargs)
-    elif internal_model == "openai-chat-completion":
+    elif internal_model == DanswerGenAIModel.OPENAI_CHAT.value:
         return OpenAIChatCompletionQA(timeout=timeout, api_key=api_key, **kwargs)
-    elif internal_model == "huggingface-inference-completion":
-        api_key = api_key if api_key is not None else GEN_AI_HUGGINGFACE_API_TOKEN
-        return HuggingFaceInferenceCompletionQA(api_key=api_key, **kwargs)
-    elif internal_model == "huggingface-inference-chat-completion":
-        api_key = api_key if api_key is not None else GEN_AI_HUGGINGFACE_API_TOKEN
-        return HuggingFaceInferenceChatCompletionQA(api_key=api_key, **kwargs)
-    # Note GPT4All is not supported for M1 Mac machines currently, removing until support is added
-    # elif internal_model == "gpt4all-completion":
-    #    return GPT4AllCompletionQA(**kwargs)
-    # elif internal_model == "gpt4all-chat-completion":
-    #    return GPT4AllChatCompletionQA(**kwargs)
+    elif internal_model == DanswerGenAIModel.GPT4ALL.value:
+        return GPT4AllCompletionQA(**kwargs)
+    elif internal_model == DanswerGenAIModel.GPT4ALL_CHAT.value:
+        return GPT4AllChatCompletionQA(**kwargs)
+    elif internal_model == DanswerGenAIModel.HUGGINGFACE.value:
+        return HuggingFaceCompletionQA(api_key=api_key, **kwargs)
+    elif internal_model == DanswerGenAIModel.HUGGINGFACE_CHAT.value:
+        return HuggingFaceChatCompletionQA(api_key=api_key, **kwargs)
+    elif internal_model == DanswerGenAIModel.REQUEST.value:
+        if endpoint is None or model_host_type is None:
+            raise ValueError(
+                "Request based GenAI model requires an endpoint and host type"
+            )
+        if model_host_type == ModelHostType.HUGGINGFACE.value:
```

Comment:
If I'm reading this right, you can't currently use hugging face with a different prompt processor besides `WeakModelFreeformProcessor`?

Reply:
Well it's not like it's configurable anyway, it's always using the default prompt_processor of the class, but I guess this could cause some confusion as this is the only case where the default parameter is changed, so maybe someone will think they can just change the default value in the class and the change is not actually reflected

Reply:
How do you propose fixing this or is this fine? I think it kinda makes sense... the downside isn't major

Reply:
Also, I cant run the 70b version anyway... cant use A100 GPUs without special request